spopt/tests/test_p_dispersion.py

```diff
@@ -0,0 +1,160 @@
+            self.facilities_snapped,
+            "geometry",
+            p_fac=2,
+        )
+        result = pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+        self.assertIsInstance(result, PDispersion)
+
+
+class TestRealWorldLocate(unittest.TestCase):
+    def setUp(self) -> None:
+        self.dirpath = os.path.join(os.path.dirname(__file__), "./data/")
+        network_distance = pandas.read_csv(
+            self.dirpath
+            + "SF_network_distance_candidateStore_16_censusTract_205_new.csv"
+        )
+
+        ntw_dist_piv = network_distance.pivot_table(
+            values="distance", index="DestinationName", columns="name"
+        )
+
+        self.cost_matrix = ntw_dist_piv.to_numpy()
+
+        facility_points = pandas.read_csv(self.dirpath + "SF_store_site_16_longlat.csv")
+
+        self.facility_points_gdf = (
+            geopandas.GeoDataFrame(
+                facility_points,
+                geometry=geopandas.points_from_xy(
+                    facility_points.long, facility_points.lat
+                ),
+            )
+            .sort_values(by=["NAME"])
+            .reset_index()
+        )
+
+        self.service_dist = 5000.0
+        self.p_facility = 4
+
+    def test_optimality_p_dispersion_from_cost_matrix(self):
+        pdispersion = PDispersion.from_cost_matrix(
+            self.cost_matrix, p_fac=self.p_facility
+        )
+        pdispersion = pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+        self.assertEqual(pdispersion.problem.status, pulp.LpStatusOptimal)
+
+    def test_infeasibility_p_dispersion_from_cost_matrix(self):
+        pdispersion = PDispersion.from_cost_matrix(self.cost_matrix, p_fac=17)
+        with self.assertRaises(RuntimeError):
+            pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+
+    def test_optimality_p_dispersion_from_geodataframe(self):
+        pdispersion = PDispersion.from_geodataframe(
+            self.facility_points_gdf,
+            "geometry",
+            p_fac=self.p_facility,
+        )
+        pdispersion = pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+        self.assertEqual(pdispersion.problem.status, pulp.LpStatusOptimal)
+
+    def test_infeasibility_p_dispersion_from_geodataframe(self):
+        pdispersion = PDispersion.from_geodataframe(
+            self.facility_points_gdf,
+            "geometry",
+            p_fac=17,
+        )
+        with self.assertRaises(RuntimeError):
+            pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+
+
+class TestErrorsWarnings(unittest.TestCase):
+    def setUp(self) -> None:
+
+        pol1 = Polygon([(0, 0), (1, 0), (1, 1)])
+        pol2 = Polygon([(0, 0), (1, 0), (1, 1), (0, 1)])
+        pol3 = Polygon([(2, 0), (3, 0), (3, 1), (2, 1)])
+        polygon_dict = {"geometry": [pol1, pol2, pol3]}
+
+        self.gdf_fac = geopandas.GeoDataFrame(polygon_dict, crs="EPSG:4326")
+
+    # error raised for diff. crs w/ x2 gdf input
+    def test_error_p_dispersion_different_crs(self):
```

Comment:
@erinrolson You are right about this. We don't need this test, because only 1 geodataframe is passed in, as you said. It was my oversight. Let go ahead and remove that test. Then I think all tests will be green!